File: core/java/src/main/java/hara/lang/protocol/IFind.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.ifind", name = "IFind")
public interface IFind<K, V> {
  @HaraMethod(value = "find", arity = 2)
  V find(K key);

  default boolean has(K key) {
    return find(key) != null;
  }
}

File: core/rust/src/lang/protocol/ifind.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.ifind", name = "IFind")]
pub trait IFind<K> {
    type Output;

    #[hara_method(value = "find", arity = 2)]
    fn find(&self, key: &K) -> Option<Self::Output>;

    fn has(&self, key: &K) -> bool {
        self.find(key).is_some()
    }
}
